### PR TITLE
Make named parameters work with :name, $name, @name and name

### LINKED ISSUE #237: Named parameters don't work
Table schema
```
CREATE TABLE IF NOT EXISTS counter(country TEXT, city TEXT, value INT, PRIMARY KEY(country, city)) WITHOUT ROWID
```
Insert/Update
```
curl -X POST -d '{"statements":[{"q":"insert into counter(country, city, value) values(\"US\", :city, 1)","params":{"city":"GDA"}}]}' https://haaawk:<password>@haaawk1-haaawk.turso.io
[{"error":{"message":"NOT NULL constraint failed: counter.city"}}]%
```
Same SQL with positional parameter works:
```
curl -X POST -d '{"statements":[{"q":"insert into counter(country, city, value) values(\"US\", ?, 1)","params":["GDA"]}]}' https://haaawk:<password>@haaawk1-haaawk.turso.io
[{"results":{"columns":[],"rows":[]}}]%
```
It seems that when using parameter name `:city` instead of `city` in params, everything works:
```
curl -X POST -d '{"statements":[{"q":"insert into counter(country, city, value) values(\"US\", :city, 1)","params":{":city":"NYC"}}]}' https://haaawk:<password>@haaawk1-haaawk.turso.io
[{"results":{"columns":[],"rows":[]}}]%
```
Same when naming parameter `@city` instead of `:city`:
```
% curl -X POST -d '{"statements":[{"q":"insert into counter(country, city, value) values(\"US\", $city, 1)","params":{"$city":"SEA"}}]}' https://haaawk:<password>@haaawk1-haaawk.turso.io
[{"results":{"columns":[],"rows":[]}}]%
```

### PR DESCRIPTION
**Pull request by haaawk**
Fixes #237

### CODE REVIEW

File: sqld/src/query.rs [hunk 1]
```diff
@@ -381,8 +381,8 @@ mod test {
         let con = rusqlite::Connection::open_in_memory().unwrap();
         let mut stmt = con.prepare("SELECT :first || $second").unwrap();
         let mut params = HashMap::new();
-        params.insert(":first".to_owned(), Value::Integer(10));
-        params.insert("$second".to_owned(), Value::Integer(20));
+        params.insert("first".to_owned(), Value::Integer(10));
+        params.insert("second".to_owned(), Value::Integer(20));
```
Comment (haaawk) [inline sqld/src/query.rs:385]: 😄

File: sqld/src/query.rs [hunk 2]
```diff
@@ -389,6 +391,19 @@ mod test {
         assert_eq!(stmt.expanded_sql().unwrap(), "SELECT 10 || 20");
     }
 
+    #[test]
+    fn test_bind_params_positional_named_no_prefix() {
+        let con = rusqlite::Connection::open_in_memory().unwrap();
+        let mut stmt = con.prepare("SELECT :first || $second").unwrap();
```
Comment (MarinPostma) [inline sqld/src/query.rs:397]: What happens if you try with a statement like 'select :first || $first'?

Comment (haaawk) [reply sqld/src/query.rs:397]: It won't allow it:
```
missing parameters, expected 2 found 1
```
here's the change:
```
diff --git a/sqld/src/query.rs b/sqld/src/query.rs
index b75bdf5..739f689 100644
--- a/sqld/src/query.rs
+++ b/sqld/src/query.rs
@@ -394,10 +394,9 @@ mod test {
     #[test]
     fn test_bind_params_positional_named_no_prefix() {
         let con = rusqlite::Connection::open_in_memory().unwrap();
-        let mut stmt = con.prepare("SELECT :first || $second").unwrap();
+        let mut stmt = con.prepare("SELECT :first || $first").unwrap();
         let mut params = HashMap::new();
         params.insert("first".to_owned(), Value::Integer(10));
-        params.insert("second".to_owned(), Value::Integer(20));
         let params = Params::new_named(params);
         params.bind(&mut stmt).unwrap();
```

Comment (haaawk) [reply sqld/src/query.rs:397]: When I worked around it by adding second parameter `$first`, it failed:
```
```
here's the change:
```
```

Comment (haaawk) [reply sqld/src/query.rs:397]: Thanks for pointing it out @MarinPostma

**Comment by haaawk** [comment]
I swapped the order of fetching named params so that $name, :name and @name will take precedence over name.
Added a test to verify that.

**Review by MarinPostma — COMMENTED** [review]
can you also update this document to reflect the changes? https://github.com/libsql/sqld/blob/main/docs/http_api.md

File: sqld/src/query.rs (hunk 2 above)
Comment (haaawk) [reply sqld/src/query.rs:400]: done.

**Comment by haaawk** [comment]
I rebased, changed the check that number of params in statement <= number of params in request to checking that each param in statement has a match in request and updated the docs.

**Review by MarinPostma — APPROVED** [review]
looks good :)

bors merge

**Comment by MarinPostma** [comment]
bors merge

**Pull request merged**